raster-foundry/raster-foundry
# Platform Routes

## Overview

This PR adds routes for platforms, organizations, and teams including authorization and validation of REST paths.

Notable changes:

  - there is now a `User.WithGroupRole` case class that adds a `groupRole` field. This is used when listing users from the context of a group and we want to communicate the role that each user has within the group.

  - the `page` method in the dao has been made generic to allow generating paged responses for composites other than the current model. For example, this is used by the `UserGroupRoleDao` to return a `PaginatedResponse[User.WithGroupRole]` from the `listUsersByGroup` method rather than just a `PaginatedResponse[UserGroupRole]`.

- validation for REST paths is done synchronously in order to leverage akka-http's `validate` method. The validation is done inside the route tree definition so that multiple endpoints can share the same validation logic. We validate that for any organization-level routes that the platform and organization id's indeed match a valid organization-platform relationship. We do the same for team routes and verify the team-organization-platform relationship.
### Checklist

~- [] Symlinks from new migrations present or corrected for any new migrations~
~- [ ] Any content changes are properly templated using `BUILDCONFIG.APP_NAME`~

### Demo

Optional. Screenshots, `curl` examples, etc.

### Notes

- there are a few unimplemented methods that will be fixed now that another PR has been merged
- the swagger spec still needs to be done
- there should be some property testing at least for the `UserGroupRoleDao.listUsersByGroup` method

## Testing Instructions

 * Using some stubbed platform/organization/user data (which is floating around in the base PR and other PRs) make requests to the new endpoints and verify they perform as expected.

Closes #3285 
Closes #3283
